### DISCUSSION TITLE
🔊 Generate provenance graphs

### DISCUSSION THREAD

**Discussion by shnizzedy** [FCP-INDI/CPAC-Development#4] [Ideas]
### Related problem

The `CpacProvenance` in our JSON sidecars is difficult to read, especially for non-experts.

### Proposed feature

For a given `CpacProvenance` (e.g.
```JSON
{
	"CpacProvenance": [
		[
			[
				"bold:func_ingress",
				"desc-preproc_bold:func_truncate",
				"desc-preproc_bold:func_despike"
			],
			[
				"TR:func_metadata_ingress"
			],
			[
				"tpattern:func_metadata_ingress"
			],
			"desc-preproc_bold:func_slice_time",
			"desc-preproc_bold:func_reorient"
		],
		[
			[
				"bold:func_ingress"
			],
			[
				[
					"bold:func_ingress",
					"desc-preproc_bold:func_truncate",
					"desc-preproc_bold:func_despike"
				],
				[
					"TR:func_metadata_ingress"
				],
				[
					"tpattern:func_metadata_ingress"
				],
				"desc-preproc_bold:func_slice_time",
				"desc-preproc_bold:func_reorient"
			],
			[
				[
					"T1w:anat_ingress",
					"desc-preproc_T1w:anatomical_init",
					"desc-preproc_T1w:n4_bias_correction",
					"desc-preproc_T1w:brain_mask_niworkflows_ants"
				],
				[
					"T1w:anat_ingress",
					"desc-preproc_T1w:anatomical_init",
					"desc-preproc_T1w:n4_bias_correction",
					"space-T1w_desc-brain_mask:brain_mask_niworkflows_ants"
				],
				"desc-brain_T1w:brain_extraction"
			],
			[
				"T1w:anat_ingress",
				"desc-preproc_T1w:anatomical_init",
				"desc-preproc_T1w:n4_bias_correction",
				"space-T1w_desc-brain_mask:brain_mask_niworkflows_ants"
			],
			"space-bold_desc-brain_mask:bold_mask_anatomical_refined"
		],
		"desc-preproc_bold:bold_masking"
	],
	"Description": "The skull-stripped BOLD time-series.",
	"SkullStripped": true,
	"Sources": [
		"desc-preproc_bold",
		"space-bold_desc-brain_mask"
	]
}
```
) generate a graph with labeled edges
![labeled edges](https://user-images.githubusercontent.com/5974438/171276373-73158725-232a-4873-87cf-81b6a07083df.png)
or linked attributes
![linked attributes](https://user-images.githubusercontent.com/5974438/171276535-ad49b269-e504-4152-9cf2-1c60ca6b9277.png)
that visualizes the provenance

### Additional context
* I hand-generated the DOT code for the example graphs above; there may be errors in there.
* Would this go in the logs? In the outputs? In the working directory? Somewhere else?
* We'd probably want this to be configurable whether these generate.

**Comment by gkiar** [comment]
I love this feature idea. I think it's an important thing to bring up in Documentation discussions moving forward.

We would want this, in my mind, in the same directory as the log files.
